### DISCUSSION TITLE
Feedback: Merges by pull request merge queue now attributed to Merge Queue bot

### DISCUSSION THREAD

**Discussion by willsmythe** [community/community#58673] [Pull Requests]
### Select Topic Area

Product Feedback

### Body

We've shipped a fix to ensure merges by the pull request merge queue are always attributed to the GitHub Merge Queue bot (`github-merge-queue[bot]`). This change applies to new merges by the queue, and will appear in the [activity view](https://github.blog/changelog/2023-05-31-view-repository-pushes-on-the-new-activity-view/) and in `push` webhook events.

> This change was made because it's possible for multiple pull requests to be merged by the queue at the same time (i.e. in the same ref update). Before this change, these ref updates were being "randomly" (not technically, but this is how it appeared) attributed to one of the users that had enqueued one of the merged pull requests. This was definitely not the correct behavior. We made the decision to attribute all merge queue "merge" ref updates to the MQ bot because it would be correct for both scenarios (single PR and multi-PR merges by the queue).

See https://github.blog/changelog/2023-06-21-merges-by-pull-request-merge-queue-now-attributed-to-the-github-merge-queue-bot/

**Comment by breml** [comment]
Hi @willsmythe 
In our CI pipelines we use `github.event.sender.login` to send a message to a release slack channel. Now all the releases are triggered by the `github-merge-queue[bot]`, which breaks our mapping for the notifications. Is the old value (the user login of the user triggering the merge of a PR) available in a different variable?

**Reply by breml** [reply]
We tried `github.event.workflow_run.head_commit.author.username` since we are already using `github.event.workflow_run.head_commit.author.name` successfully, but this did not work.

**Reply by breml** [reply]
@willsmythe after some more testing, we found, that in the complete `github` context, the only two references to the original author of the commit are `github.event.workflow_run.head_commit.author.email` and `github.event.workflow_run.head_commit.author.name`, which are both not suitable / reliable for our purposes we do no longer have a way to find the original author during the deployment phase in our github action pipeline.

**Reply by danvendia** [reply]
We are running into the same issue as well :/

**Reply by JavaScriptBach** [reply]
Same here. This change degrades the usefulness of our Slack notifications for no benefit. @willsmythe can we revert this change, or have a way to restore the old behavior?

**Reply by willsmythe** [reply]
Thanks for the feedback everyone (and sorry for the inconvenience). We're discussing options.

For more context ...

This change was made because it's possible for multiple pull requests to be merged by the queue at the same time (i.e. in the same ref update). Before this change, these ref updates were being "randomly" (not technically, but this is how it appeared) attributed to one of the users that had enqueued one of the merged pull requests. This was definitely not the correct behavior. We made the decision to attribute all merge queue "merge" ref updates to the MQ bot because it would be correct for both scenarios (single PR and multi-PR merges by the queue).

With that said, we're looking at options.

Independent of this, `pull_request` is the best event to trigger workflows about a specific PR. When a PR is merged, an event with `action` of `closed` is fired. The `pull_request.merged_by` field will tell you who merged it (a `null` value indicates the PR was closed with being merged).

Similarly, a `pull_request` event with `action` `dequeued` is published when a PR is removed from the queue for any reason, including because it was merged. The `reason` field indicates why the PR was removed. For example:

```json
{
  "action": "dequeued",
  "number": 334,
  "pull_request": {
   "merged_by": {
      "login": "willsmythe",
    },
  },
  "reason": "MERGE"
}
```

Regardless of the number of pull requests merged by the queue at the same time, `pull_request` webhook events are published for each pull request. The same is not true for `push` events (only a single event is published when the base branch is updated).

**Reply by ejholmes** [reply]
Just to pile on here, this has broken author attribution in third party tools that we use, like Terraform Enterprise/Cloud. I get there's technical reasons behind this (the possibility of two changes getting collapsed into one), but it feels like that's probably a semi-rare occurrence and this optimizes in the wrong direction.

We're much more interested in knowing who actually authored the commits, than what tool/process merged it. I think picking one of the authors is better.

**Reply by jcw-** [reply]
This has broken our notification tooling as well - hoping for a revert...

**Reply by breml** [reply]
We solved this issue by adding the following additional step to our pipeline:

```
      - uses: octokit/request-action@v2.x
        id: get_head_commit
        with:
          route: GET /repos/${{ github.repository }}/commits/${{ github.event.workflow_run.head_commit.id }}
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

The login of the github user who authored the last commit is then available through: `${{ fromJson(steps.get_head_commit.outputs.data).author.login }}`.

**Reply by abolla22** [reply]
Im trying the same to get the github author . Tried with this Action but keep getting this error any idea whats causing it @breml  ? 
Error seeing: 
GET /repos/${{ github.repository }} /commits/github_id_fix
> mediaType: [object Object]
< 404

**Reply by breml** [reply]
No, not from the details you provided.

**Reply by mtamiola-priv** [reply]
have you try to provide commit sha instead of branch name @abolla22 ?

**Comment by promiseofcake** [comment]
This is more of a breaking change than a fix in my opinion.

**Comment by phobologic** [comment]
+1 this is causing issues with systems that really need proper attribution.  Would love a fix here, thanks.

**Comment by dylanklohr** [comment]
I'm looking to possibly introduce merge queues as part of an updated deploy strategy at my company. For those being impacted by this particular change, are commits that are being attributed to `github-merge-queue[bot]` _only_ attributed to the bot, or are they showing as "co-authored"?

**Reply by ejholmes** [reply]
The actual git commit data (like if you were to `git pull`, or look at the commit in the GH UI) contains the correct author information. Just webhooks + GitHub API shows `github-merge-queue[bot]`.

**Comment by pww217** [comment]
This has been problematic for our CI too. We write helm values into a protected branch as part of the CI, and previously we had used a bot account that we owned and allowed it to bypass PR requirements to push these commits. I still want to require it for all devs and other actions, though.

But since turning on merge queues, the `github.actor` is `github-merge-queue[bot]` which is an implicit account that doesn't exist among my contributors to the repo, which means I can't add them to the PR check exception like I could for the old bot. I even tried to add the numerical account ID but no dice.

The weirdest part, though, is that sometimes the CI just works anyway. 3/4 times it fails ands says "lacking permissions to skip a PR", but sometimes it just works, and I can't figure out why so far.

There's extremely little info about this online, which is not surprising because Merge Queues are pretty new. Unfortunately for now we may just need to disable it for now because I don't see a good solution if I can't allow `github-merge-queue[bot]` to bypass PR checks or basically be treated like a normal Github account.

Would appreciate any advice, insights, or workarounds folks have of course!

**Comment by kudojp** [comment]
This caused a problem in the repository in my organization.

## Issue explained

- I enabled merge queue onto `main` branch.
- I configured `.circleci/config.yml` as below so that the job `deploy_onto_staging` uses a context in it.

```yaml
  - deploy_onto_staging:
    when:
      and:
        - equal: [<< pipeline.git.branch >>, main]
    context:
      - our_internal_context # This causes a problem.
    ...
```

In this way,

- ⭕ When I merge a PR to `main` branch manually, this job succeeded. This is because `our_internal_context` is configured to be accessible by `All members` in my organization.
- ❌ When a PR is merged with merge queue, this job failed. This is because `github-merge-queue[bot]` does not have an access to `our_internal_context`.

## Possible solution

`github-merge-queue[bot]` should have the same access privilege to contexts as a user who enqueued the PR to the merge queue.

**Comment by kleisauke** [comment]
Sorry for replying to an old discussion, but I think this change may have broken the closing keywords in commit messages. According to [GitHub's documentation](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword), commit messages can include closing keywords:

> You can also use closing keywords in a commit message. The issue will be closed when you merge the commit into the default branch, but the pull request that contains the commit will not be listed as a linked pull request.

However, when commits are merged via a merge queue, the issue does not automatically close. For example, issue https://github.com/libjxl/libjxl/issues/4525 was fixed via PR https://github.com/libjxl/libjxl/pull/4577, but the issue remained open after the commit (https://github.com/libjxl/libjxl/commit/ae5cb195d491208c94a082871adab805af6cd81e) landed in the default branch. I assume `github-merge-queue[bot]` does not have sufficient permissions to close an issue?
